### PR TITLE
Flexible callable

### PR DESCRIPTION
**Pull request by sixolet**
This is a draft implementation of the work discussed in https://github.com/python/typing/issues/264

Most of the subtlety was in getting the function subtyping rules (what I believe to be) right.

Unfortunately basically all the typeshed stub files violate these rules all over the place -- specifically, arguments in a function overriding a superclass function must be named the same thing, otherwise calling that function by keyword is going to fail.  So I still need to resolve that stuff.

Also I need to write a pile more tests.

Also also apparently I haven't implemented StarArg and KwArg yet.  

Actually there's a decent amount left to do.

### CODE REVIEW

File: mypy/messages.py [hunk 1]
```diff
@@ -167,8 +167,40 @@ def format(self, typ: Type, verbosity: int = 0) -> str:
...
+                for arg_name, arg_type, arg_kind in zip(
+                        func.arg_names, func.arg_types, func.arg_kinds):
+                    if arg_kind == ARG_POS and arg_name is None or verbosity == 0:
+                        arg_strings.append(
+                            strip_quotes(
+                                self.format(
+                                    arg_type,
+                                    verbosity = max(verbosity-1, 0))))
+                    else:
+                        constructor = {
```
Comment (sixolet) [inline mypy/messages.py]: TODO: Move this constant dict thing out to a constant

File: mypy/messages.py [hunk 2]
```diff
@@ -167,8 +167,40 @@ def format(self, typ: Type, verbosity: int = 0) -> str:
...
+                            ARG_POS: "Arg",
+                            ARG_OPT: "DefaultArg",
+                            ARG_NAMED: "Arg",
+                            ARG_NAMED_OPT: "DefaultArg",
+                            ARG_STAR: "StarArg",
+                            ARG_STAR2: "KwArg",
+                        }[arg_kind]
+                        if arg_kind in (
```
Comment (sixolet) [inline mypy/messages.py]: Move tuple constant out to a constant.

File: extensions/mypy_extensions.py [hunk 3]
```diff
@@ -20,3 +22,15 @@ def new_dict(*args, **kwargs):
...
+    def __init__(name=None, typ=Any, keyword_only=False):
+        self.name = name
+        self.typ = typ
+        self.named_only = named_only
+
+class DefaultArg(object):
+    def __init__(name=None, typ=Any, keyword_only=False):
+        self.name = name
+        self.typ = typ
+        self.named_only = named_only
```
Comment (sixolet) [inline extensions/mypy_extensions.py]: Add the rest of the kinds of arguments.

File: mypy/parse.py [hunk 4]
```diff
@@ -821,7 +839,12 @@ def construct_function_type(self, args: List[Argument], ret_type: Type,
         if ret_type is None:
             ret_type = AnyType(implicit=True)
         arg_kinds = [arg.kind for arg in args]
-        arg_names = [arg.variable.name() for arg in args]
+        if include_names:
+            arg_names = [arg.variable.name()
```
Comment (sixolet) [inline mypy/parse.py]: This implements "names starting with __ are unnamed arguments".  It's probably necessary to actually get the typeshed files passing.

File: mypy/subtypes.py [hunk 5]
```diff
@@ -269,46 +268,138 @@ def is_callable_subtype(left: CallableType, right: CallableType,
...
-    """Is left a subtype of right, assuming left has *args?
+        # Get possible corresponding arguments by name and by position.
+        if not done_with_positional:
+            left_by_position = left.argument_by_position(i)
+        else:
+            left_by_position = None
+        if right_name is not None:
+            left_by_name = left.argument_by_name(right_name)
+        else:
+            left_by_name = None
```
Comment (sixolet) [inline mypy/subtypes.py]: Blank lines totally match each other!  For sure!

File: mypy/visitor.py [hunk 6]
```diff
@@ -22,220 +22,220 @@ class NodeVisitor(Generic[T]):
 
     # Module structure
```
Comment (sixolet) [inline mypy/visitor.py]: These are all  overloaded later with different arg names.  See also needing the positional-only syntax.

File: test-data/unit/check-classes.test [hunk 7]
```diff
@@ -1230,9 +1230,9 @@ class D:
     def __getattr__(self, x: str) -> None: pass
 [out]
 main: note: In member "__getattr__" of class "B":
-main:4: error: Invalid signature "def (self: __main__.B, x: __main__.A) -> __main__.B"
```
Comment (sixolet) [inline test-data/unit/check-classes.test]: I removed argument names from special functions so these error messages changed.

File: test-data/unit/check-functions.test [hunk 8]
```diff
@@ -1275,6 +1275,67 @@ f('x') # E: Argument 1 to "f" has incompatible type "str"; expected "int"
...
+int_str_fun = isf_unnamed
+int_named_str_fun = isf_unnamed # E: Incompatible types in assignment (expression has type Callable[[int, str], str], variable has type Callable[[int, Arg('s', str, False)], str])
+int_opt_str_fun = iosf
+int_str_fun = iosf
+int_opt_str_fun = isf # E: Incompatible types in assignment (expression has type Callable[[Arg('ii', int, False), Arg('ss', str, False)], str], variable has type Callable[[int, DefaultArg('None', str, False)], str])
+
+int_named_str_fun = isf # E: Incompatible types in assignment (expression has type Callable[[Arg('ii', int, False), Arg('ss', str, False)], str], variable has type Callable[[int, Arg('s', str, False)], str])
+int_named_str_fun = iosf
+
+[builtins fixtures/dict.pyi]
```
Comment (sixolet) [inline test-data/unit/check-functions.test]: So many more to come.

**Comment by JukkaL** [comment]
Mypy doesn't check argument name compatibility in method overrides by design -- it would generate a huge number of errors for a lot of real-world code. Generally code like this implicitly expects callers not to use keyword arguments. Clearly this is not sound, which is unfortunate. On the other hand, users have generally not been complaining about this.

Also, many methods implemented in C only accept positional arguments. For them, we'd need a way of specifying in stub files that the arguments don't actually have names, and tracking this down for every single C method would be a pretty big task. When we previously discussed this, one idea was to use `__argname` to mean a positional-only argument.

I think that argument name compatibility checking for overrides should be left out of this PR and discussed separately, as it's a much bigger change than making `Callable` more flexible, and it would require major typeshed rework. What about having two kinds of compatibility checking for functions, one of which doesn't check argument name compatibility (except for keyword-only arguments)? We'd use the more lenient variant for checking method overrides. This way we could remain compatible with existing practice, including typeshed.

**Comment by sixolet** [comment]
@JukkaL How about I split this into a few different diffs:
 - Allow `__argname` to produce a positional-only argument (easy)
 - Implement the difference between required and optional keyword-only arguments
 - Function compatibility checks argument names, but not for positional arguments in situations where you're checking override compatibility.  It checks all of them for assignments.
 - The part where we actually parse the new `Callable` syntax and use it.

And submit them each as separate pull requests?

They might have a dependency tree among them.

**Comment by JukkaL** [comment]
Sounds like a good plan!

**Comment by sixolet** [comment]
I have started the series of smaller pull requests that this turned into.

**Pull request closed without merging**